spec/Data_Parallelism.tex

```diff
@@ -280,9 +284,110 @@ \section{Forall Intents}
 \end{rationale}
 
 
+\section{Task-Private Variables}
+\label{Task_Private_Variables}
+\index{task-private variables}
+\index{shadow variables}
+\index{data parallelism!task-private variables}
+\index{data parallelism!shadow variables}
+\index{statements!forall@\chpl{forall}!task-private variables}
+\index{statements!forall@\chpl{forall}!shadow variables}
+
+A \emph{task-private variable} declared in a forall loop results
+in a separate shadow variable in each task created by the forall
+loop's parallel iterator, as well as a separate shadow variable
+created at the top level of the parallel iterator itself.
+
+The shadow variable is created at the start of the task or the
+iterator and destroyed at the end of the task/iterator.
+Within the lexical scope of the forall statement or expression,
+the variable name references the shadow variable in the task
+that executed the current yield statement, or in the iterator
+if the current yield occurs in the iterator outside any task constructs.
+In contrast to regular forall intents \rsec{Forall_Intents},
+these shadow variables are unrelated to outer variables of the same name,
+if any.
+
+The syntax of a task-private variable declaration in a forall statement's
+with-clause is:
+
+\begin{syntax}
+task-private-var-decl:
+  task-private-var-kind identifier type-part[OPT] initialization-part[OPT]
+
+task-private-var-kind:
+  `const'
+  `var'
+  `ref'
+\end{syntax}
+
+The declaration of a \chpl{const} or \chpl{var} task-private variable must
+have at least one of \sntx{type-part} and \sntx{initialization-part}.
+A \chpl{ref} task-private variable must have \sntx{initialization-part}
+and cannot have \sntx{type-part}. A \chpl{ref} shadow variable
+is a reference to the \sntx{initialization-part} as calculated at
+the start of the corresponding task or the iterator.
+\chpl{ref} shadow variables are never destroyed.
+
+\begin{craychapel}
+Currently task-private variables are not available for task constructs.
+A regular variable declared at the start of the begin/cobegin/coforall
+block can be used instead.
+\end{craychapel}
+
+\begin{chapelexample}{task-private-variable.chpl}
+In the following example, the \chpl{writeln()} statement will observe
+the first shadow variable 4 times: twice each for the yields
+{\tt "before coforall"} and {\tt "after coforall"}.
+An additional shadow variable will be created and observed twice
+for each of the three \chpl{coforall} tasks.
+\begin{chapel}
+var cnt: atomic int;                     // count our shadow variables
```

Comment:
Should we print ``cnt`` at the end?

Reply:
I don't feel it would illustrate anything.